File: cpp/problem414.cpp
#include <mystd.h>
using namespace MYSTD;
using namespace std;

class Solution {
public:
    int thirdMax(vector<int>& nums) {
        long x1 = LONG_MIN+2, x2 = LONG_MIN+1, x3 = LONG_MIN;
        for(auto x: nums){
            if( x > x1 ){ x3 = x2; x2 = x1; x1 = x; }
            else if( x == x1 ) continue;
            else if( x > x2 ){ x3 = x2; x2 = x; }
            else if( x == x2 ) continue;
            else if( x > x3 ) x3 = x;
        }
        return x3 < INT_MIN?x1:x3;
    }  
};

int main(int argc, char const *argv[]){
    Solution sol;

    vector<int> nums;

    // cases in Leetcode Problem
    nums = {3,2,1};
    cout << sol.thirdMax(nums) << endl;
    nums = {1,2};
    cout << sol.thirdMax(nums) << endl;
    nums = {2,2,3,1};
    cout << sol.thirdMax(nums) << endl;

    // other extreme cases
    nums = {2,1,2};
    cout << sol.thirdMax(nums) << endl;
    nums = {2,2,3,3};
    cout << sol.thirdMax(nums) << endl;
    nums = {2,2,3,3,1};
    cout << sol.thirdMax(nums) << endl;
    return 0;
}

File: swift/problem414.swift
class Solution {
    func thirdMax(_ nums: [Int]) -> Int {
        var first = nums[0]; var second: Int?; var third: Int?
        
        for number in nums {
            if (number > first) {
                if (second != nil) {
                    third = second
                }
                second = first
                first = number
            } else if (number == first) {
                continue
            } else if (number > second ?? Int.min) {
                if (second != nil) {
                    third = second
                }
                second = number
            } else if (number == second) {
                continue
            } else if (number > third ?? Int.min){
                third = number
            }
        }
        
        return third ?? first
    }
}
